File: util/encodeArticles.js
// You must run npm install before this can run

var createTamp = require('../');
var _ = require('lodash');

var articles = require('./articles.json');

packIt(articles);

function packIt(articles){
	articles = articles.filter(function(article){
		return article.byline && article.byline.original;
	});

	articles.forEach(function(article, index){
		article.id = index;
	});

	var possibilities = {
		sectionName: articles.map(function(a){
			return a.section_name;
		}).filter(function(a, index, list){
			return a && !~list.indexOf(a, index + 1);
		}).sort(),
		byline: articles.map(function(a){
			return a.byline.original;
		}).filter(function(a, index, list){
			return !~list.indexOf(a, index + 1);
		}).sort()
	}

	var tamp = createTamp();

	tamp.addAttribute({
		attrName: 'byline',
		possibilities: possibilities.byline,
		maxChoices: 1
	});

	tamp.addAttribute({
		attrName: 'section_name',
		possibilities: possibilities.sectionName,
		maxChoices: 1
	});

	tamp.pack(articles.map(function(a){
		return {
			id: a.id,
			section_name: a.section_name,
			byline: a.byline.original
		}
	}));


	console.log(tamp.toJSON());
}

File: util/encode_articles.rb
# You must gem install json and tamper before this can run

require 'json'
require 'tamper'

articlesJson = File.read('articles.json')
articles = JSON.parse(articlesJson)

# Kill invalid bylines.
articles.delete_if { |a| !a['byline'].is_a?(Hash) }

# Assign each a numeric ID.  If we were storing these in a database,
# we could assign an autoincrement id.  For this demo we'll assign a
# temporary id.
#
# A numeric id is required when generating a Tamper pack.
articles.each_with_index { |a, idx| a.merge!(id: idx)}

# Generate the distinct possibilities for each attributes.
# If we were storing these in a db, we could use a DISTINCT query.
possibilities = {
  section_name: articles.map { |a| a['section_name'] }.compact.uniq.sort,
  byline: articles.map { |a| a['byline']['original'] }.compact.uniq.sort
}

# Configure pack
pack = Tamper::PackSet.new

pack.add_attribute(
  attr_name: :byline,
  possibilities: possibilities[:byline],
  max_choices: 1
)

pack.add_attribute(
  attr_name: :section_name,
  possibilities: possibilities[:section_name],
  max_choices: 1
)

pack.build_pack(num_items: articles.length, max_guid: articles.last[:id]) do |pack|
  articles.each do |a|
    pack << {
      id: a[:id],
      section_name: a['section_name'],
      byline: a['byline']['original']
    }
  end
end

puts pack.to_json
